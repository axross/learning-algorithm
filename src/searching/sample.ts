export const sortedCharactors = [
  "a",
  "b",
  "c",
  "d",
  "e",
  "f",
  "g",
  "h",
  "i",
  "j",
  "k",
  "l",
  "m",
  "n",
  "o",
  "p",
  "q",
  "r",
  "s",
  "t",
  "u",
  "v",
  "w",
  "x",
  "y",
  "z"
];

export const unsortedCharactors = [
  "c",
  "n",
  "r",
  "j",
  "m",
  "s",
  "b",
  "u",
  "w",
  "y",
  "h",
  "t",
  "v",
  "f",
  "z",
  "g",
  "x",
  "o",
  "k",
  "q",
  "d",
  "e",
  "p",
  "l",
  "a",
  "i"
];
