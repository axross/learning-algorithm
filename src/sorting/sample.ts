import roundRobin from "../test-utility/round-robin.ts";
import randomInteger from "../test-utility/random-integer.ts";

export const charactors = ["a", "b", "c", "d", "e", "f", "g", "h"];

export const samples = roundRobin(charactors).slice(1);

export function getRandomSample() {
  return samples[randomInteger(0, samples.length - 1)];
}

export const staticSample = ["f", "a", "h", "b", "d", "g", "e", "c"];

export const charactorPairs = ["a0", "a1", "b0", "b1", "c0", "c1", "d0", "d1"];

export const charactorPairSamples = roundRobin(["a0", "b0", "c0", "d0"]).reduce(
  (arr, pattern0) => [
    ...arr,
    ...roundRobin(["a1", "b1", "c1", "d1"]).reduce(
      (arr, pattern1) => [...arr, [...pattern0, ...pattern1]],
      [] as string[][]
    )
  ],
  [] as string[][]
);

export const randomCharactorPairSample =
  charactorPairSamples[randomInteger(0, charactorPairSamples.length - 1)];

export const staticCharactorPairSample =
  randomCharactorPairSample[Math.floor((samples.length / 7) * 5)];
