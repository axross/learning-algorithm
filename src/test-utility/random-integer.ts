const randomInteger = (min: number, max: number) =>
  Math.floor(Math.random() * (max - min)) + min;

export default randomInteger;
