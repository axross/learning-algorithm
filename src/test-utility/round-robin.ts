function roundRobin<Element>(
  elements: Element[],
  list?: Element[][]
): Element[][] {
  if (list === undefined) {
    list = elements.map(element => [element]);
  }

  if (list[0].length === elements.length) {
    return list;
  }

  return list.reduce(
    (arr, item) => {
      const newItems = elements
        .filter(element => !item.includes(element))
        .map(element => [...item, element]);

      return [...arr, ...roundRobin(elements, [...newItems])];
    },
    [] as Element[][]
  );
}

export default roundRobin;
