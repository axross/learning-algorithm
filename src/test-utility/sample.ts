export const sampleCharactors = [
  ...Array.from({ length: 10 }, (_, i) => String.fromCharCode(i + 48)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 65)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97))
];

export const shuffledSampleCharactors = [...sampleCharactors];

for (let i = shuffledSampleCharactors.length - 1; i >= 0; i--) {
  const rand = Math.floor(Math.random() * (i + 1));

  [shuffledSampleCharactors[i], shuffledSampleCharactors[rand]] = [
    shuffledSampleCharactors[rand],
    shuffledSampleCharactors[i]
  ];
}
