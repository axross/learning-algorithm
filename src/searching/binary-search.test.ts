import binarySearch, { BinarySearchStep } from "./binary-search";
import { sampleCharactors } from "../test-utility/sample";

describe("binarySearch", () => {
  const list = sampleCharactors;

  test(`binarySearch(target: "X") finds the index of the target to be ${sampleCharactors.indexOf(
    "X"
  )}`, () => {
    expect(
      binarySearch({
        list: list,
        target: "X",
        compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0)
      })
    ).toBe(sampleCharactors.indexOf("X"));
  });

  test("binarySearch(target: 25000) doesn't find it", () => {
    expect(
      binarySearch({
        list: list,
        target: "🍣",
        compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0)
      })
    ).toBe(-1);
  });

  test('the step snapshot by binarySearch(target: "X") matches with the previous one', () => {
    const steps: BinarySearchStep<string>[] = [];

    binarySearch({
      list: list,
      target: "X",
      compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0),
      onStep: step => steps.push(step)
    });

    expect(steps).toMatchSnapshot();
  });

  test("the maximum caliculation size is O(log_2 n + 1)", () => {
    let maximumStepLength = 0;

    for (let i = 0; i < list.length; ++i) {
      const steps: BinarySearchStep<string>[] = [];

      binarySearch({
        list: list,
        target: list[i],
        compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0),
        onStep: step => steps.push(step)
      });

      if (steps.length > maximumStepLength) {
        maximumStepLength = steps.length;
      }
    }

    expect(maximumStepLength).toBe(Math.floor(Math.log2(list.length)) + 1);
  });

  test("the avarage caliculation size is O(log_2 n)", () => {
    let totalStepLengths = 0;

    for (let i = 0; i < list.length; ++i) {
      const steps: BinarySearchStep<string>[] = [];

      binarySearch({
        list: list,
        target: list[i],
        compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0),
        onStep: step => steps.push(step)
      });

      totalStepLengths = totalStepLengths + steps.length;
    }

    expect(Math.floor(totalStepLengths / list.length)).toBe(
      Math.floor(Math.log2(list.length))
    );
  });
});
