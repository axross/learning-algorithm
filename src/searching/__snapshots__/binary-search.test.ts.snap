// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`binarySearch the step snapshot by binarySearch(target: "X") matches with the previous one 1`] = `
Array [
  Object {
    "middle": Object {
      "index": 30,
      "value": "U",
    },
    "searchRange": Object {
      "end": 61,
      "start": 0,
    },
  },
  Object {
    "middle": Object {
      "index": 46,
      "value": "k",
    },
    "searchRange": Object {
      "end": 61,
      "start": 31,
    },
  },
  Object {
    "middle": Object {
      "index": 38,
      "value": "c",
    },
    "searchRange": Object {
      "end": 45,
      "start": 31,
    },
  },
  Object {
    "middle": Object {
      "index": 34,
      "value": "Y",
    },
    "searchRange": Object {
      "end": 37,
      "start": 31,
    },
  },
  Object {
    "middle": Object {
      "index": 32,
      "value": "W",
    },
    "searchRange": Object {
      "end": 33,
      "start": 31,
    },
  },
  Object {
    "middle": Object {
      "index": 33,
      "value": "X",
    },
    "searchRange": Object {
      "end": 33,
      "start": 33,
    },
  },
]
`;
